### DISCUSSION TITLE
Perhaps we should add module destructors as well?

### DISCUSSION THREAD

**Discussion by cheong00** [dotnet/csharplang#4635] [Language Ideas] [answered]
Since we will likely have [module initializers](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-9.0/module-initializers.md) on C#9, and the rule don't forbid the use of unmanaged resources in there, perhaps it make sense to also add "module destructors" for teardown?

**Comment by huoyaoyuan** [comment]
Module initializer wasn't added, but exposed from C#. It requires IL having the feature first.

**Comment by sharwell** [comment]
AppDomain.Unload is probably the closest we have to a module destructor. It has proven extremely unreliable in practice. If a module needs to run code during application shutdown, it needs to be a normal part of code flow and cannot be a user-defined finalizer. Otherwise in practice you'll end up with 25% success, 25% silently fails to run, 25% deadlock that leaves the process running in the background, and 25% crash.

**Comment by Joe4evr** [comment, marked as answer]
I believe this is relevant here: https://devblogs.microsoft.com/oldnewthing/20120105-00/?p=8683
> Don’t worry about freeing memory; it will all go away when the process address space is destroyed. Don’t worry about closing handles; handles are closed automatically when the process handle table is destroyed.

**Reply by cheong00** [reply]
No, I'm not worry about memory or handles be stored as SafeHandle, but things that is not bookkept by the system such as device states of hardware (say, DCB for serial ports), or kernel handles that opened by 3rd party unmanaged DLLs, or unmanaged hardware libraries with documentation that explicitly require cleanup calls (say, the DLL for automation of HikVision security cameras).

**Reply by qrli** [reply]
> but things that is not bookkept by the system such as device states of hardware (say, DCB for serial ports), or kernel handles that opened by 3rd party unmanaged DLLs, or unmanaged hardware libraries with documentation that explicitly require cleanup calls (say, the DLL for automation of HikVision security cameras).

That sounds like exactly what SafeHandle (with underlying critical finalizer) was designed for.

**Reply by cheong00** [reply]
For the part about DLL cleanup, it's to be called when you no longer use the DLL. In the case of HikVision library it's to be called after you cleanup all handles to cameras but before DLL is unloaded.

Also note that I'm talking about using unmanaged DLL that calls other unmanaged DLLs, so things like SafeHandle are not relevant.

**Reply by qrli** [reply]
For unmanaged DLL, the support is even weaker, as unmanaged world can play a lot of tricks and they do. E.g. some other software (e.g. anti-virus) may hook the DLL API and change the timing and order things happened.

As @huoyaoyuan mentioned, it is a limitation of .NET runtime rather than C#. But I can kind of imagine why such support was not added, from the history of various issues happened with `DLL_PROCESS_DETACH`. It is very hard to do right in practice (most things no longer work there), similar to finalizer. Even finalizers have been discouraged for long, it is better not to let applications to play with DLL unload hook. 

But you are also right that there are cases it is indeed necessary. Then it boils down to RoI for such feature. Without such a feature, we can still make an explicit normal function call to clean up, which is more controllable on timing and threading. It may not be bad at all.

**Reply by Joe4evr** [reply]
If I understand the way `SafeHandle`s work correctly, wouldn't [something kinda like this](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4AEBMBGAsAKBQBgAIUMA6AJQFcA7YGAWwlIGEB7egBxgBtoBlaGBgBjCAGcA3LgKEAsiKisxrOMFIB1GNQDMaUnwCGcCAAkD1ACa9JuXFuDRqB7oTERnEC8TRzWFyrwAIgbABoQghIbGZpa88tSUYgDy1BAAkmJp1GDOMF4A9PmsUIQA7gAWIRCQJQDWWl4qkUam5lYQhAislISpnrgA3riEI4QA2rJ+AenUMHS5AF7QALrDo/aOzsQYAGzEACyEWXMAFACUhAC8AHyEAPpaYqHUooQA/G+XvRClvv5BIQM5ykODWIw4sByDm2e0m/wgwVC90ez1EILBhAhMChHTh00RQLOGNG4UIACMDG4TqxStQxDF2hFgFBKBAiTgSUMOaMAL62bkjFDaQisGqwCwdMmsVguDJZHLcPJXW7AcqKX6pX4AOVYwDSnF4jFongAogAPUQcOisajA/kkiG6iDCByNMV5SXSlzkCC8SmtWIQc4Yrkkkmq9XfbW6/UcQ0QY0Wc2W622s4gkl8nBZoA) work to accomplish this from the managed side?

**Reply by cheong00** [reply]
Okay, it should work. Just not something that newer users of the language possibly able to figure out immediately.

**Reply by Joe4evr** [reply]
Well, newer users to the language shouldn't be directly using advanced interop features like SafeHandles and ModuleInitializers, anyway. 🙃

**Reply by jnm2** [reply]
> Okay, it should work. Just not something that newer users of the language possibly able to figure out immediately.

@BillWagner Is this helpful to track for potential docs?
